retrieval/discovery/kubernetes/discovery.go

```diff
@@ -735,3 +788,233 @@ func nodeHostIP(node *Node) (net.IP, error) {
+	until(func() {
+		pods, resourceVersion, err := kd.getPods()
+		if err != nil {
+			log.Errorf("Cannot initialize pods collection: %s", err)
+			return
+		}
+		kd.podsMu.Lock()
+		kd.pods = pods
+		kd.podsMu.Unlock()
+
+		req, err := http.NewRequest("GET", podsURL, nil)
+		if err != nil {
+			log.Errorf("Cannot create pods request: %s", err)
+			return
+		}
+
+		values := req.URL.Query()
+		values.Add("watch", "true")
+		values.Add("resourceVersion", resourceVersion)
+		req.URL.RawQuery = values.Encode()
+		res, err := kd.queryAPIServerReq(req)
+		if err != nil {
+			log.Errorf("Failed to watch pods: %s", err)
+			return
+		}
+		defer res.Body.Close()
+		if res.StatusCode != http.StatusOK {
+			log.Errorf("Failed to watch pods: %d", res.StatusCode)
+			return
+		}
+
+		d := json.NewDecoder(res.Body)
+
+		for {
+			var event podEvent
+			if err := d.Decode(&event); err != nil {
+				log.Errorf("Watch pods unexpectedly closed: %s", err)
+				return
+			}
+
+			select {
+			case events <- &event:
+			case <-done:
+			}
+		}
+	}, retryInterval, done)
+}
+
+func podSource(pod *Pod) string {
+	return sourcePodPrefix + ":" + pod.ObjectMeta.Namespace + ":" + pod.ObjectMeta.Name
+}
+
+type ByContainerPort []ContainerPort
+
+func (a ByContainerPort) Len() int           { return len(a) }
+func (a ByContainerPort) Less(i, j int) bool { return a[i].ContainerPort < a[j].ContainerPort }
+func (a ByContainerPort) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+
+func updatePodTargets(pod *Pod, allContainers bool) []model.LabelSet {
+	var targets []model.LabelSet = make([]model.LabelSet, 0, len(pod.PodSpec.Containers))
+	if pod.PodStatus.PodIP == "" {
+		log.Debugf("skipping pod %s -- PodStatus.PodIP is empty", pod.ObjectMeta.Name)
+		return targets
+	}
+
+	if pod.PodStatus.Phase != "Running" {
+		log.Debugf("skipping pod %s -- status is not `Running`", pod.ObjectMeta.Name)
+		return targets
+	}
+
+	ready := "unknown"
+	for _, cond := range pod.PodStatus.Conditions {
+		if strings.ToLower(cond.Type) == "ready" {
+			ready = strings.ToLower(cond.Status)
+		}
+	}
+
+	for _, container := range pod.PodSpec.Containers {
+		// Collect a list of TCP ports
+		// Sort by port number, ascending
+		// Product a target pointed at the first port
+		// Include a label containing all ports (portName=port,PortName=port,...,)
+		var tcpPorts []ContainerPort
+		var portLabel bytes.Buffer
+
+		for _, port := range container.Ports {
+			if port.Protocol == "TCP" {
+				tcpPorts = append(tcpPorts, port)
+			}
+		}
+
+		if len(tcpPorts) == 0 {
+			log.Debugf("skipping container %s with no TCP ports", container.Name)
+			continue
+		}
+
+		sort.Sort(ByContainerPort(tcpPorts))
+
+		for _, port := range tcpPorts {
+			portLabel.WriteString(port.Name)
+			portLabel.WriteString("=")
+			portLabel.WriteString(strconv.FormatInt(int64(port.ContainerPort), 10))
+			portLabel.WriteString(",")
```

Comment:
For clarity, it'd be best to move the `portLabel` definition to just before this `for`.